neo/Consensus/ConsensusService.cs

```diff
@@ -231,22 +322,168 @@ private void OnPrepareRequestReceived(ConsensusPayload payload, PrepareRequest m
+            /// Thus, we need to remove the signature from the Payload to correctly verify Speaker identity agreements with this block
+            /// </summary>
+            byte[] tempSignature = message.PrepReqSignature;
+            message.PrepReqSignature = new byte[64];
+            payload.Data = message.ToArray();
+
+            if (!Crypto.Default.VerifySignature(payload.GetHashData(), tempSignature, context.Validators[payload.ValidatorIndex].EncodePoint(false)))
+                return false;
+
+            /// <summary>
+            /// These next 2 lines could be removed, because payload is not anymore used
+            /// it was already saved before changed in the context.PreparePayload... However, let keep things clean for now
+            /// </summary>
+            message.PrepReqSignature = tempSignature;
+            payload.Data = message.ToArray();
+            return true;
+        }
+
+
         private void OnPrepareResponseReceived(ConsensusPayload payload, PrepareResponse message)
         {
-            if (context.Signatures[payload.ValidatorIndex] != null) return;
+            if (context.State.HasFlag(ConsensusState.CommitSent)) return;
+            /// <summary>
+            /// This payload.ValidatorIndex already submitted a not null signature
+            /// </summary>
+            if (context.SignedPayloads[payload.ValidatorIndex] != null) return;
+
             Log($"{nameof(OnPrepareResponseReceived)}: height={payload.BlockIndex} view={message.ViewNumber} index={payload.ValidatorIndex}");
-            byte[] hashData = context.MakeHeader()?.GetHashData();
-            if (hashData == null)
+
+            /// <summary>
+            /// ***** The following is like an additional feature ******
+            /// ORIGINAL CODE: In the original code we were just storing the signature and later verifying when the Payload really arrives.
+            /// FEATURED ONE: Here we can check the PreparePayload if it did not arrive, because all PrepareREsponse Payloads carries that
+            /// </summary>
+            if (context.PreparePayload == null)
             {
-                context.Signatures[payload.ValidatorIndex] = message.Signature;
+                Log($"{nameof(OnPrepareRequestReceived)}: Response before PrepRequest, trying to speed up p2p route...");
+                if (message.PreparePayload.ValidatorIndex != context.PrimaryIndex) return;
+                if (!CheckPrimaryPayloadSignature(message.PreparePayload)) return;
+                Log($"{nameof(OnPrepareRequestReceived)}: indirectly from index={payload.ValidatorIndex}");
+                OnPrepareRequestReceived(message.PreparePayload, context.GetPrepareRequestMessage(message.PreparePayload));
             }
-            else if (Crypto.Default.VerifySignature(hashData, message.Signature, context.Validators[payload.ValidatorIndex].EncodePoint(false)))
+
+            /// <summary>
```

Comment:
`summary` should be used for classes and methods, no for regular comments ( I never seen this before, so i think that is not a standard, maybe i'm wrong )

Reply:
AHUehauehaueahueahuea
You and @erikzhang told me this once and I started to use for everything on C#. aehuaheuahueaea

We gonna remove comments after all tests and correct all this.

Reply:
https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/xmldoc/summary

"The <summary> tag should be used to describe a type or a type member. Use <remarks> to add supplemental information to a type description. Use the cref Attribute to enable documentation tools such as Sandcastle to create internal hyperlinks to documentation pages for code elements."

A lot of tricks